Gunakan data titanic dari seaborn.  
* target=survived  
* fitur= 'pclass','sex','age','fare'  



* isi missing value  
* Cek multicollinearity
* split data menjadi train dan test (80:20)
* buat intepretasi hasil summary
* cek akurasi model terhadap data test



### Import Library

In [61]:
import statsmodels.api as sm
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import normaltest
import statsmodels.api as sm

### Dataset

In [26]:
titanic=sns.load_dataset('titanic')
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

### Missing Value

In [27]:
titanic.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [28]:
titanic['age'].fillna(titanic['age'].mean(),
inplace=True)
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [29]:
titanic.isna().sum()

survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Dummy Variable

In [34]:
titanic_dummy=pd.get_dummies(titanic, drop_first=True) #yang pertama muncul akan dihilangkan
titanic_dummy

survived  pclass        age  sibsp  parch     fare  adult_male  alone  \
0           0       3  22.000000      1      0   7.2500        True  False   
1           1       1  38.000000      1      0  71.2833       False  False   
2           1       3  26.000000      0      0   7.9250       False   True   
3           1       1  35.000000      1      0  53.1000       False  False   
4           0       3  35.000000      0      0   8.0500        True   True   
..        ...     ...        ...    ...    ...      ...         ...    ...   
886         0       2  27.000000      0      0  13.0000        True   True   
887         1       1  19.000000      0      0  30.0000       False   True   
888         0       3  29.699118      1      2  23.4500       False  False   
889         1       1  26.000000      0      0  30.0000        True   True   
890         0       3  32.000000      0      0   7.7500        True   True   

     sex_male  embarked_Q  ...  who_woman  deck_B  deck_C  deck_D  deck_E  \
0           1           0  ...          0       0       0       0       0   
1           0           0  ...          1       0       1       0       0   
2           0           0  ...          1       0       0       0       0   
3           0           0  ...          1       0       1       0       0   
4           1           0  ...          0       0       0       0       0   
..        ...         ...  ...        ...     ...     ...     ...     ...   
886         1           0  ...          0       0       0       0       0   
887         0           0  ...          1       1       0       0       0   
888         0           0  ...          1       0       0       0       0   
889         1           0  ...          0       0       1       0       0   
890         1           1  ...          0       0       0       0       0   

     deck_F  deck_G  embark_town_Queenstown  embark_town_Southampton  \
0         0       0                       0                        1   
1         0       0                       0                        0   
2         0       0                       0                        1   
3         0       0                       0                        1   
4         0       0                       0                        1   
..      ...     ...                     ...                      ...   
886       0       0                       0                        1   
887       0       0                       0                        1   
888       0       0                       0                        1   
889       0       0                       0                        0   
890       0       0                       1                        0   

     alive_yes  
0            0  
1            1  
2            1  
3            1  
4            0  
..         ...  
886          0  
887          1  
888          0  
889          1  
890          0  

[891 rows x 24 columns]

In [90]:
titanic_dummy.columns

Index(['survived', 'pclass', 'age', 'sibsp', 'parch', 'fare', 'adult_male',
       'alone', 'sex_male', 'embarked_Q', 'embarked_S', 'class_Second',
       'class_Third', 'who_man', 'who_woman', 'deck_B', 'deck_C', 'deck_D',
       'deck_E', 'deck_F', 'deck_G', 'embark_town_Queenstown',
       'embark_town_Southampton', 'alive_yes'],
      dtype='object')

### Model

In [102]:
fitur=['pclass','sex_male','age','fare']
target='survived'
sm_logit=sm.Logit(y,sm.add_constant(x))
result=sm_logit.fit()

Optimization terminated successfully.
         Current function value: 0.451877
         Iterations 6


In [103]:
print(result.summary())

                           Logit Regression Results                           
Dep. Variable:               survived   No. Observations:                  891
Model:                          Logit   Df Residuals:                      886
Method:                           MLE   Df Model:                            4
Date:                Thu, 01 Jul 2021   Pseudo R-squ.:                  0.3214
Time:                        13:45:14   Log-Likelihood:                -402.62
converged:                       True   LL-Null:                       -593.33
Covariance Type:            nonrobust   LLR p-value:                 2.888e-81
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          4.6642      0.509      9.166      0.000       3.667       5.662
pclass        -1.1501      0.135     -8.504      0.000      -1.415      -0.885
sex_male      -2.6071      0.187    -13.915      0.0

In [99]:
x=titanic_dummy[fitur]
y=titanic_dummy[target]

In [104]:
x.describe()

pclass    sex_male         age        fare
count  891.000000  891.000000  891.000000  891.000000
mean     2.308642    0.647587   29.699118   32.204208
std      0.836071    0.477990   13.002015   49.693429
min      1.000000    0.000000    0.420000    0.000000
25%      2.000000    0.000000   22.000000    7.910400
50%      3.000000    1.000000   29.699118   14.454200
75%      3.000000    1.000000   35.000000   31.000000
max      3.000000    1.000000   80.000000  512.329200

### Colinearity Check

In [106]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
def calc_vif(X):
    # calculating vif
    vif=pd.DataFrame()
    vif['variables']=X.columns
    vif['VIF']=[variance_inflation_factor(X.values,i) for i in range (X.shape[1])]
    return vif

In [107]:
calc_vif(x)

variables       VIF
0    pclass  3.769757
1  sex_male  2.903523
2       age  4.126846
3      fare  1.425912

### Model Intepretation

In [111]:
#B1 (pclass)
print(np.exp(-1.1501*1)) 
np.exp(1.1501*1) 

0.31660510728524643


3.1585087447722273

In [110]:
#B2 (sex)
print(np.exp(-2.6071*1))
np.exp(2.6071*1)

0.07374810345184897


13.559670733131625

In [112]:
#B3 (age)
print(np.exp(-0.0332*1))
np.exp(0.0332*1)

0.9673450712265954


1.0337572700215427

In [114]:
#B4 (fare)

np.exp(0.0006*1)

#tidak dilanjutkan dihitung karena non signifikan

1.0006001800360054

### Summary

**LLR-test (LLR p-value):**

B1, B2, B3, B4: p-value <=0.05, artinya kita punya cukup bukti bahwa ada setidaknya satu fitur yang memiliki pengaruh signifikan terhadap kemungkinan survived.

**Wald Test(P>|z|):**
1. B0 (constant): p-val <= 0.05, artinya kita butuh B0 di model kita
2. B1 (pclass): p-val <= 0.05, artinya kita punya cukup bukti bahwa pclass mempengaruhi nilai kemungkinan survived 
3. B2 (sex=who_man): p-val <= 0.05, artinya kita punya cukup bukti bahwa laki-laki kemungkinan survived 
4. B3 (age): p-val <= 0.05, artinya bahwa age mempengaruhi nilai kemungkinan survived
5. B4 (fare): p-val > 0.05 artinya, tidak cukup bukti bahwa fare mempengaruhi kemungkinan survived

**Parameter Estimate (coef)**
1. B1=  -1.1501: pclass menurunkan kemungkinan survived. ketika class menurun 1 tingkat dan variabel lainnya tetap, maka kemungkinan survived menurun 3.15 kali 
2. B2=  -2.6071 : Ratio sex pria dan survived menurunkan kemungkinan tenggelam. ketika jumlah pria dengan nilai survived menurun, maka kemungkinan tenggelam menurun 13.55 kali 
3. B3=   -0.0332: umur menurunkan kemungkinan survived. ketika umur menurun 1 tingkat, maka kemungkinan survived menurun 1 kali
4. B4= 0.0006: tidak diinterprestasikan karena hasil uji tidak signifikan


**Coefficient Determination (Pseudo R-squ)**

Pseudo R-squared: 32.14%, Model yang kita buat mampu menjelaskan 32.14% dari varian default rate atau varian kemungkinan gagal selamat.

### Validation

Train data= 80%

Test data= 20%

In [115]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [116]:
x_train, x_test, y_train, y_test= train_test_split(
    x,
    y,
    stratify=y, #membagi secara acak, ratio antar kelas di data train dan test sama
    test_size=0.2,
    random_state=2020
)

In [117]:
sm_logit=sm.Logit(y_train,sm.add_constant(x_train))
result_train=sm_logit.fit()

Optimization terminated successfully.
         Current function value: 0.448174
         Iterations 6


In [118]:
y_predict_proba=result_train.predict(sm.add_constant(x_test))
y_predict_proba

560    0.083861
130    0.075220
551    0.237366
587    0.237608
2      0.588478
         ...   
818    0.053649
113    0.640144
605    0.068876
642    0.777540
206    0.078701
Length: 179, dtype: float64

In [119]:
y_predict_class=np.where(y_predict_proba > .5,1,0) #1 tidak selamat, 0 selamat
y_predict_class

array([0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1,
       0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1,
       0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0,
       1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 0])

In [120]:
print('model accuracy score in test dataset:', accuracy_score(y_test, y_predict_class))

model accuracy score in test dataset: 0.7932960893854749


Akurasi 79.88%, artinya model yang kita buat mampu memprediksi secara tepat 7 dari 10 kali